.github/workflows/test.yaml

```diff
@@ -223,6 +212,10 @@ jobs:
               export SKIP_GOFMT=1
               echo "Formatting checks will be skipped due to the use of Go version ${{ matrix.gochannel }}"
           fi
+          if [ "${{ matrix.gochannel }}" != "latest/stable" ]; then
+              export SKIP_GOLANG_CI_LINT=1
+              echo "Will skip golangci_lint checks on all but the latest/stable version"
```

Comment:
Ah good idea, I was curious about whether we should parse the golangci-lint version and make sure it's compatible with the go version, but it's probably fine to assume that latest/stable golangci-lint will work with latest/stable go.

I'm not entirely certain though, since the golangci-lint snap is 3rd party (by @alexmurray) it's possible it could lag behind the go version. Bumps in go versions have been known to break golangci-lint, as occurred recently when go 1.23 reached stable: https://github.com/canonical/snapd/pull/14359

Reply:
On second thought, I'm curious/worried if "advancements" to the `golangci-lint` snap and go could get more strict and report existing code as bad. I know there's a `--new-from-rev` arg used when run in CI, but not when run independently. So I'd maybe prefer to run on go `1.18/stable` instead/as well?

Reply:
For your second comment, I could set the GITHUB_BASE_REF default to "remotes/upstream/master" (that will then be overridden in the CI to be just "master") so that by default run-checks.sh should only check new code just as it does in the CI. Would that help? Without a default, to mimic the CI behavior, one would run `GITHUB_BASE_REF=remotes/upstream/master ./run-checks --static`. (This assuming the person running the run-checks.sh forked snapd and set the upstream)

Reply:
For your first comment, do you think the only safe way we can use a snap for golangci-lint is to maintain that snap ourselves?

Reply:
FWIW I'd be happy to say transfer this snap to the Canonical publisher account and add various snapd folks as collaborators if that makes this easier?

Reply:
> For your second comment, I could set the GITHUB_BASE_REF default to "remotes/upstream/master" (that will then be overridden in the CI to be just "master") so that by default run-checks.sh should only check new code just as it does in the CI. Would that help? Without a default, to mimic the CI behavior, one would run `GITHUB_BASE_REF=remotes/upstream/master ./run-checks --static`. (This assuming the person running the run-checks.sh forked snapd and set the upstream)

I don't think we can rely on people having set an `upstream` remote. I'd just as soon incrementally enforce new linting restrictions on the code (including existing code) as we bump golangci-lint versions. I just would like to have snapd and the golangci-lint snaps synchronized so when there's an update to golangci-lint, we know and can make the adjustments to snapd at the same time.

From what I can tell (please @alexmurray correct me if I'm wrong) golangci-lint compatibility changes in two ways:
1. There are specific versions of the golangci-lint source code which check for different lints, generally getting more strict over time
2. The version of golang with which golangci-lint is compiled governs the compatibility with particular go versions

With those assumptions, it seems like for snapd, we want to lock in (1) and do a version bump periodically (once a year?). But we want/need to keep (2) up to date with the `latest/stable` branch of the `go` snap, since that's what we use when we run static checks on `latest/stable` in CI. So it seems like what we need is something like an automatic rebuild of golangci-lint built on the `latest/candidate` channel of `go`. Ideally, we'd have at least a day of any new go version in `latest/candidate` before it's moved to `latest/stable`, giving time for `golangci-lint` to be rebuilt before snapd starts using `go` `latest/stable` in CI. However, this wasn't the case for 1.23 :upside_down_face: 
```
channels:
  latest/stable:       1.23.0       2024-08-14 (10698)  68MB classic
  latest/candidate:    ↑                                     
  latest/beta:         ↑                                     
  latest/edge:         1.24-e705a2d 2024-08-07 (10683) 111MB classic
  1.23/stable:         1.23.0       2024-08-14 (10698)  68MB classic
  1.23/candidate:      1.23.0       2024-08-14 (10698)  68MB classic
  1.23/beta:           ↑                                     
  1.23/edge:           ↑                                     
```

So perhaps it would make sense to move golangci-lint under Canonical and add something like a `latest/stable/snapd` channel which is pinned to a particular `golangci-lint` version, but is rebuilt with the `go` `latest/candidate` snap? I suggest this rather than having branches for particular golangci-lint versions, since then we presumably have to maintain any old version we move away from or else explicitly close that channel.

Reply:
Then again maybe I'm way overthinking this, and really we should just track latest/stable `golangci-lint` (and ensure it's built on the `go` snap `latest/stable` or `latest/candidate` channel) and occasionally snapd CI breaks and we need to fix it immediately. This would be much simpler, and would force us to keep up to date, at least in our local branches. Since we're using `GITHUB_BASE_REF` in CI (but presumably not locally) we shouldn't break CI more broadly, just PR branches, which we'll want to comply with the latest constraints anyway. Perhaps this should simply be called `BASE_REF` rather than `GITHUB_BASE_REF`, so that it's clearer that local devs can set the variable to restrict the set of commits over which `golangci-lint` is run locally, if for example their local run fails and they want to see whether it's the result of their new changes or existing code.

Reply:
Just to summarize, at least for now, simply using @alexmurray's snap is fine for snapd's purposes because 1) it's built with go from latest/stable and 2) he updates it regularly. 

Since the nightly build doesn't use golangci-lint and our CI uses `--new-from-rev` (which developers can also use by populating the environment variable `BASE_REF` before running `run-checks.sh`), golangci-lint should only find issues in new code and leave the old code in peace.

Assuming I didn't miss anything, all the issues in this thread should be resolved.